src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -195,6 +188,11 @@ void VMThread::run() {
   // _thread_in_native state to block.
   VM_Exit::wait_for_threads_in_native_to_block();
 
+  // The ObjectMonitor subsystem uses perf counters so do this before
+  // we signal that the VM thread is gone. We don't want to run afoul
+  // of perfMemory_exit() in exit_globals().
+  ObjectSynchronizer::do_final_audit_and_print_stats();
```

Comment:
Again I wonder why this was moved so far from the original call to deflate? It makes sense to do the final deflation once the final safepoint has been reached, but should we do it first once the safepoint has been reached?

Reply:
Again the old code made an async deflation request which meant
that the ServiceThread would eventually do the work, but that was
in a racy fashion which is what motivated this change.

The `exit_globals()` call is a little less obvious for this code path.
That happens over in `src/hotspot/share/runtime/thread.cpp` in the
`Threads::destroy_vm()` function. In `destroy_vm()` we wait for the
VMThread to exit via a `VMThread::wait_for_vm_thread_exit()` call
and a bit later we call `exit_globals()`. 

In the new code, we're calling `do_final_audit_and_print_stats()` as
part of the VMThread's exit path code in the `VMThread::run()`
function so, again, we're doing the final audit a little bit earlier than
we used to do it before.